### PR TITLE
add GaussianScaleMixture distribution

### PR DESCRIPTION
**Pull request by martinjankowiak**
partner to #1189

### CODE REVIEW

**Comment by jpchen** [comment]
nit: is there a reason you call them Normal in some distributions and Gaussians in others?

**Comment by martinjankowiak** [comment]
by default i always use normal. except in this case Gaussian Scale Mixture is a well known term from the literature

e.g. https://papers.nips.cc/paper/1750-scale-mixtures-of-gaussians-and-the-statistics-of-natural-images.pdf

File: pyro/distributions/gaussian_scale_mixture.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            coeffs[k + 1:] *= self.dim - 2 * (k + 1)
+        return coeffs
+
+    def log_prob(self, value):
+        assert value.dim() == 1 and value.size(0) == self.dim
+        epsilon_sqr = torch.pow(value / self.scale, 2.0).sum()
+        lambdas_power = torch.pow(self.lambdas, -self.dim)
+        result = lambdas_power * self.categorical.probs * \
+            torch.exp(-0.5 * epsilon_sqr / torch.pow(self.lambdas, 2.0))  # K
+        result = torch.log(result.sum())
```
Comment (fritzo) [inline pyro/distributions/gaussian_scale_mixture.py]: nit: I'm surprised you didn't use `log_sum_exp()` here

File: pyro/distributions/gaussian_scale_mixture.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        g = grad_output  # l i
+        g = g.unsqueeze(-2)  # l 1 i
+
+        lambdas_sqr = torch.pow(lambdas, 2.0)  # j
+        epsilons = z / scale  # l i
+        epsilons_sqr = torch.pow(epsilons, 2.0)  # l i
+        r_sqr = epsilons_sqr.sum(-1)  # l
+        r_sqr_j = r_sqr.unsqueeze(-1) / lambdas_sqr  # l j
+        log_scale = torch.log(scale)  # i
+        scale_product = log_scale.sum().exp()
```
Comment (fritzo) [inline pyro/distributions/gaussian_scale_mixture.py]: nit: `scale_product = scale.prod()`

File: pyro/distributions/gaussian_scale_mixture.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    def backward(ctx, grad_output):
+        z, scale, logits, lambdas, pis, coeffs = ctx.saved_tensors
+        dim = scale.size(0)
+        g = grad_output  # l i
+        g = g.unsqueeze(-2)  # l 1 i
+
+        lambdas_sqr = torch.pow(lambdas, 2.0)  # j
+        epsilons = z / scale  # l i
+        epsilons_sqr = torch.pow(epsilons, 2.0)  # l i
+        r_sqr = epsilons_sqr.sum(-1)  # l
```
Comment (fritzo) [inline pyro/distributions/gaussian_scale_mixture.py]: nit: `epsilons_sqr.sum(-1, keepdim=True)`

File: pyro/distributions/gaussian_scale_mixture.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        r_sqr_j = r_sqr.unsqueeze(-1) / lambdas_sqr  # l j
+        log_scale = torch.log(scale)  # i
+        scale_product = log_scale.sum().exp()
+        lambdas_power = torch.pow(lambdas, float(dim))
+
+        q_j = torch.exp(-0.5 * r_sqr_j) / math.pow(2.0 * math.pi, 0.5 * float(dim))  # l j
+        q_j /= scale_product * lambdas_power  # l j
+        q_tot = (pis * q_j).sum(-1)  # l
```
Comment (fritzo) [inline pyro/distributions/gaussian_scale_mixture.py]: nit: `.sum(-1, keepdim=True)`

File: tests/distributions/test_gaussian_mixtures.py [hunk 5]
```diff
@@ -5,11 +5,11 @@
...
 import pytest
-from pyro.distributions import MixtureOfDiagNormalsSharedCovariance
+from pyro.distributions import MixtureOfDiagNormalsSharedCovariance, GaussianScaleMixture
 from tests.common import assert_equal
 
 
-@pytest.mark.parametrize('mix_dist', [MixtureOfDiagNormalsSharedCovariance])
+@pytest.mark.parametrize('mix_dist', [MixtureOfDiagNormalsSharedCovariance, GaussianScaleMixture])
 @pytest.mark.parametrize('K', [3])
 @pytest.mark.parametrize('D', [2, 4])
```
Comment (fritzo) [inline tests/distributions/test_gaussian_mixtures.py]: Can you add `D=1` example to test your `dim > 1` conditional logic?

File: pyro/distributions/gaussian_scale_mixture.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    [1] Pathwise Derivatives for Multivariate Distributions, Martin Jankowiak &
+    Theofanis Karaletsos. arXiv:1806.01856
+
+    Note that this distribution supports both even and odd dimensions, but the
+    former should be more a bit higher precision, since it doesn't use any erfs in
+    the backward call.
+
+    :param torch.tensor scale: D-dimensional vector of scale
+    :param torch.tensor logits: K-dimensional vector of logits
+    :param torch.tensor lambdas: K-dimensional matrix of sigma multipliers
```
Comment (fritzo) [inline pyro/distributions/gaussian_scale_mixture.py]: We generally try to prefer semantic names over greek. Consider renaming to something like
```py
coord_scale
component_scale
component_logits
```

File: pyro/distributions/gaussian_scale_mixture.py (hunk 1 above)
Comment (martinjankowiak) [reply pyro/distributions/gaussian_scale_mixture.py]: left TODO to do this once torch.logsumexp is in pytorch release

File: pyro/distributions/gaussian_scale_mixture.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/distributions/gaussian_scale_mixture.py]: done

File: pyro/distributions/gaussian_scale_mixture.py (hunk 3 above)
Comment (martinjankowiak) [reply pyro/distributions/gaussian_scale_mixture.py]: done

File: pyro/distributions/gaussian_scale_mixture.py (hunk 4 above)
Comment (martinjankowiak) [reply pyro/distributions/gaussian_scale_mixture.py]: done

File: tests/distributions/test_gaussian_mixtures.py (hunk 5 above)
Comment (martinjankowiak) [reply tests/distributions/test_gaussian_mixtures.py]: added D=1 assert error

File: pyro/distributions/gaussian_scale_mixture.py (hunk 6 above)
Comment (martinjankowiak) [reply pyro/distributions/gaussian_scale_mixture.py]: done

**Comment by martinjankowiak** [comment]
re: TransformedDistribution approach, one could do this but i think it would be slower? (also unwinding the math would be a huge pain)

File: pyro/distributions/diag_normal_mixture_shared_cov.py [hunk 7]
```diff
@@ -18,92 +18,99 @@ class MixtureOfDiagNormalsSharedCovariance(TorchDistribution):
...
+            assert(component_logits.size(-1) == locs.size(-2))
             batch_shape = tuple(locs.shape[:-2])
         self.locs = locs
-        self.scale = scale
-        self.logits = logits
+        self.coord_scale = coord_scale
+        self.component_logits = component_logits
         self.dim = locs.size(-1)
-        self.categorical = Categorical(logits=logits)
+        assert (self.dim > 1), "This distribution does not support dimension D = 1"
```
Comment (fritzo) [inline pyro/distributions/diag_normal_mixture_shared_cov.py]: Can you raise a `NotImplementedError` here?
```py
if self.dim < 2:
    raise NotImplementedError('This distribution does not support D = 1')
```
Then you might get a mug.

Comment (martinjankowiak) [reply pyro/distributions/diag_normal_mixture_shared_cov.py]: done. @jpchen pleaaaasssse go get me a mug!?

File: pyro/distributions/gaussian_scale_mixture.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    :param torch.tensor component_logits: K-dimensional vector of logits
+    :param torch.tensor component_scale: K-dimensional vector of scale multipliers
+    """
+    has_rsample = True
+    arg_constraints = {"component_scale": constraints.positive, "coord_scale": constraints.positive,
+                       "component_logits": constraints.real}
+
+    def __init__(self, coord_scale, component_logits, component_scale):
+        self.dim = coord_scale.size(0)
+	if self.dim < 2:
```
Comment (fritzo) [inline pyro/distributions/gaussian_scale_mixture.py]: What happened to the indentation here?

Comment (fritzo) [reply pyro/distributions/gaussian_scale_mixture.py]: mug
